Updates for the release process (#3376)
grafana/k6: .github/ISSUE_TEMPLATE/release.md

### At the beginning of the cycle

> - [ ] Get from `@k6-browser` team the Release coordinator on their side and co-assign the issue.

Reviewer: ```suggestion
- [ ] Obtain the Release coordinator's contact from the `@k6-browser` team and co-assign the issue to them.
```

- [ ] Create a new `release-notes-v0.4x.0` long-lived branch and add a new release notes file using the available [template](/grafana/k6/tree/master/release%20notes/template.md) to the [repository's `release notes` folder](/grafana/k6/tree/master/release%20notes).
- [ ] Go through the potential [dependencies updates](Dependencies.md) and create a dedicated PR if any of them is relevant to this release.

> - [ ] Create a new `release-v0.4x.0` long-lived branch on the [k6-docs](https://github.com/grafana/k6-docs) repository. Try to keep the branch synced with `main` during the cycle for not getting an outdated branch and get a lot of conflicts.

Reviewer: ```suggestion
- [ ] Create a new `release-v0.4x.0` long-lived branch on the [k6-docs](https://github.com/grafana/k6-docs) repository. Try to keep it synchronized with `main` throughout the development cycle to prevent it from becoming outdated and to avoid numerous conflicts.
```

> - [ ] Create a new `release-v0.4x.0` long-lived branch on the [k6-docs](https://github.com/grafana/k6-docs) repository. Try to keep it synchronized with `main` throughout the development cycle to prevent it from becoming outdated and to avoid numerous conflicts.

Reviewer 2: I think that this will be problematic.

The reason why (at least I) have made this PR in the last few days and requested people to rebase on top of it (or not at all), is specifically because we do changes to the documentation over time.

And this PR archives the previous version - copying the current JS API (I think this is the only versioned part) to a separate folder. 

If you do this *early* in the cycle that will mean that you will get the API that are at the beginning.

So you either have to:
1. Not do it at the beginning - but then you can not merge changes for the next version in as you will want to archive without them
2. do it at the beginning and somehow every time there is update there - update the archived version. 
3. something else

We have been doing 3 - where we just do it at the very end and then the changes (which are *mostly* in those API JS) go in on top of it.

There will not really be any *conflicts* as you just copied the code before hand, so unless a small fix at the current version and a new JS API for a future verison collide - you won't even have indication this is a problem.

Reviewer 3: My understanding is that we're not doing archiving at the beginning of the cycle, but creating a blank branch that will contain only documentation of upcoming features. It's just about that rebasing the single branch on top of one that contains the archived API is easier that multiply feature branches.

Reviewer 3: An assumption based on fact that `1 day before release` contains: 

```
- [ ] Open a PR from `main` in the `k6-docs` repository, archiving the current k6's JavaScript API version as per the following [instructions](https://github.com/grafana/k6-docs/wiki/Add-version-for-Javascript-API-documentation).
```

Author: As Oleg said, the intent here is to merge on this long-lived branch only the new documentation without archiving. The day before the release we open the PR for archiving **directly from master**, we merge this new PR and then merge this long-lived branch.

Reviewer 2: Okay, but then my worry is that this will basically mean that keeping the history of this long-lived branch will be a nightmare especially once you have collisions.

And with more changes you have bigger chances for collisions over and over again.

So I am not certain how this is an improvement to have multiple *specific* PRs for the specific changes open that we merge on the day of the release - as we have done so far.

If somebody (k6-browser team seems to run in this)  has changes that build on top of on another - they can have their own branch.

But IME most of the times we have updates to different parts of the API that do not interact.

Author: I don't have a strong opinion here because I see the potential nightmare regarding the sync of the branch. On the other hand, as the one who did the last release, I didn't like the process of following and managing multiple PRs for the docs.

@inancgumus do you have any input from your team? Would be equally fine for you if we merge the PRs only at the end and you continue to use the current strategy?

Reviewer 4: We have used long-lived branches on k6-docs for the past two cycles and never had a merging problem, even with a 2K LOC branch, updated by many throughout the cycles. It might be because we only change browser-specific parts. This approach made our workflow much more uncomplicated. But your mileage may vary. If you don't follow this approach, we can still open browser-specific long-lived branches and merge them with the ones you open at the beginning or the end of a cycle (before a k6 release). @ka3de Thoughts?

Reviewer 5: I think this summarizes it pretty well:

> If you don't follow this approach, we can still open browser-specific long-lived branches and merge them near the end of a cycle with the ones you open (whether you open them at the beginning or the end of a cycle). But, if you follow this approach, there will be no sync delay between the core and browser teams.

Our aim is to reduce the required syncs between browser and core team as much as possible during the k6 release. For example avoiding having _N_ PRs open pending to be merged during the release process for docs and TS Type definition updates. As @inancgumus mentioned, this can be kept as an internal procedure for us, for which we create long-lived "cycle" PRs for browser docs/ts types/release notes to which we add incremental changes during the cycle, and then during the release day we have to sync with the core team to merge them. Or if the core team already has these "long-lived" cycle PRs, then we can merge directly our changes during the cycle development so there is no need for explicit sync during the release day (which sometimes results painful, conflicts are generated due to PR merges etc), and instead, whatever is in these branches is what will be included in the release.

Author: @olegbespalov you are the next one in the chain, do you have a preference? I would like to make an attempt for the next cycle following the flow we already described in this PR. But in case of a lot of conflicts will appear then it would impact you, so I would prefer to leave to you the final approval.

Reviewer 3: I don't mind trying the approach that we've just described :shrug: 

In any case, I don't see all of these as set in stone, I believe even in our checklist we fixed cyclish reconsidering if something doesn't work well :sweat_smile:

- [ ] Ensure all PRs in the k6 repository, part of the current [milestone](https://github.com/grafana/k6/milestones), have been merged.
- [ ] Open a PR with the release notes for the new version, and ask teams who might have contributed to the release (@k6-browser, @k6-chaos, @devrel teams, etc.) to contribute their notes and review the existing ones.
- [ ] Share the release notes PR with the k6 open-source teams. Request contributions from all affected teams (browser, chaos, devrel, docs, etc.) and any other stakeholders involved in the new release.
- [ ] Open a separate PR for bumping [the k6 Go project's version](https://github.com/grafana/k6/blob/9fa50b2d1f259cdccff5cc7bc18a236d31c345ac/lib/consts/consts.go#L11).
- [ ] Open a PR in the `DefinitelyTyped/DefinitelyTyped` repository, using the release branch created in the grafana/k6-DefinitelyTyped fork, to update the k6 type definitions for the new release.

#### ~ 1 day before the release date.

- [ ] Open a PR from `main` in the `k6-docs` repository, archiving the current k6's JavaScript API version as per the following [instructions](https://github.com/grafana/k6-docs/wiki/Add-version-for-Javascript-API-documentation).

> - [ ] Open a PR in `k6-docs` repository for the release branch containing all the merged PRs for the upcoming version. If any, resolve the conflicts with `main` branch.

Reviewer: ```suggestion
- [ ] Open a PR in `k6-docs` repository for the release branch containing all the merged PRs for the upcoming version. If any, resolve the conflicts with the `main` branch.
```

### Release Day

#### Documentation

- [ ] Merge the `k6-docs` repository's Javascript API archiving PR and rebase the rest of the branches meant for the release on top of the new state of the `main` branch.

> - [ ] Merge the release PR for `k6-docs` containing all the changes related to the release. The order with the the previous PR is relevant otherwise the new changes will be included in the previous versions.

Reviewer: ```suggestion
- [ ] Merge the release PR for `k6-docs` containing all the changes related to the release. The order in which this PR is merged relative to the previous one is crucial; otherwise, the new changes might inadvertently be incorporated into previous versions.
```
